doc/cprover-manual/contracts-assigns.md

```diff
@@ -1,50 +1,166 @@
 [CPROVER Manual TOC](../../)
 
-# Assigns Clause
+# Assigns Clauses
 
-## In Function Contracts
+An _assigns clause_ lets the user to specify which memory locations may be
+assigned to by a function or a loop.
 
-### Syntax
+For a function contract, if no assigns clause is provided, the default set of
+locations is the empty set.
+
+For a loop contract, if no assigns clause is provided with the contract, CBMC
+attempts to infer the set of locations assigned by the loop from the loop body,
+and then checks that this set is correct as if it had been provided by the user.
+
+For both functions and loop contracts, if more than one assigns clause is
+provided, they are implicitly merged into a single clause.
+
+Each target listed in the clause gets resolved to a pair of values of type
+`(void *, size_t)` specifying the assignable memory location's start address
+and extent in bytes.
+
+We use the _assigns_ interpretation for these locations, as opposed to the
+_modifies_ interpretation. This means that memory locations that are part of the
+_frame_ of the function call or the loop that are not listed in the assigns
+clause (or in the inferred clause for a loop contract) must not be assigned to
+by the function or the loop, even if they end up holding the same value as they
+held before the function call or before entering the loop.
+
+In contrast, memory location that are locally stack- or heap-allocated during
+function execution or loop execution are by definition not part of the _frame_
+and can always be assigned to by the function or the loop.
+
+## Syntax
 
 ```c
-__CPROVER_assigns(*identifier*, ...)
+__CPROVER_assigns( *targets* )
+```
+
+Where targets have the following syntax:
+
+```
+targets ::= conditional-target-group (';' conditional-target-group)* ';'?
+conditional-target-group ::= (condition ':')? target (',' target)*
+target ::= l-value-expression
+        | __CPROVER_typed_target(l-value-expression)
+        | __CPROVER_object_whole(pointer-typed-expression)
+        | __CPROVER_object_from(pointer-typed-expression)
+        | __CPROVER_object_upto(pointer-typed-expression,
+                                unsigned-integer-typed-expression)
+```
+
+
+For function contracts, the condition and target expressions
+in the assigns clause may only involve function parameters,
+global variables or type identifiers in `sizeof` or cast expressions.
+The target expression must be free of function calls and side-effects.
+The condition expression may contain calls to side-effect-free functions.
+
+For a loop contract, the condition and target expressions in the assigns clause
+may involve any identifier that is in scope at loop entry
+(parameters of the surrounding function, local variables, global variables,
+type identifiers in `sizeof` or cast expressions, etc.).
+The target expression must be free of function calls and side-effects.
+The condition expression may contain calls to side-effect-free functions.
+
+### Lvalue targets
```

Comment:
We need to show what would be equivalent now to say `__CPROVER_assign(*ptr)`.

Reply:
+1

Reply:
`__CPROVER_assigns(*ptr)` can still be used.

Reply:
But that is the only one we don't explain the semantics in the text.

Reply:
`*ptr` is an lvalue so the definition is given in the section:

> An l-value target `expr` with a complete type `expr_t` specifies that the range of `sizeof(expr_t)` bytes starting at `&expr` is assignable.

Reply:
Added examples of lvalue targets vs non-lvalue targets.